File: leetcode/go/product_except_self.go
package main

import (
	"fmt"
	"reflect"
)

func productExceptSelf_2nSpace(nums []int) []int {
	n := len(nums)
	fromHead := make([]int, n)
	fromTail := make([]int, n)
	fromHead[0], fromTail[n - 1] = 1, 1
	for i := 1; i < n; i++ {
		fromHead[i] = fromHead[i - 1] * nums[i - 1]
		fromTail[n - i - 1] = fromTail[n - i] * nums[n - i]
	}
	result := make([]int, n)
	for i := 0; i < n; i++ {
		result[i] = fromHead[i] * fromTail[i]
	}
	return result
}

func productExceptSelf(nums []int) []int {
	n := len(nums)
	result := make([]int, n)
	for i := 0; i < n; i++ {
		result[i] = 1
	}
	fromHead, fromTail := 1, 1
	for i := 1; i < n; i++ {
		fromHead *= nums[i - 1]
		result[i] *= fromHead
		fromTail *= nums[n - i]
		result[n - i - 1] *= fromTail
	}
	return result
}

type TestData struct {
	input    []int
	expected []int
}

func main() {
	datas := []TestData{{[]int{1, 2, 3, 4}, []int{24, 12, 8, 6}}}
	for index, data := range datas {
		fmt.Printf("Test Round #%d\n", index)
		fmt.Println("Input nums:", data.input)
		result2nSpace := productExceptSelf_2nSpace(data.input)
		fmt.Println("Output result of 2N space algorithm:",
			result2nSpace)
		fmt.Println("Expected ?", reflect.DeepEqual(result2nSpace,
			data.expected))
		fmt.Println("------------------------------------------")
		result := productExceptSelf(data.input)
		fmt.Println("Output result of N space algorithm:", result)
		fmt.Println("Expected ?", reflect.DeepEqual(result,
			data.expected))
		fmt.Println()
	}
}

File: leetcode/product_except_self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """

        sz = len(nums)
        if sz <= 1:
            return []

        output = [1] * sz
        
        for i in range(1, sz):
            output[i] = nums[i - 1] * output[i - 1]

        tmp = 1
        for i in range(sz - 2, -1, -1):
            output[i] = nums[i + 1] * output[i] * tmp
            tmp *= nums[i + 1]
            print(output, tmp)

        return output



if __name__ == '__main__':
    cases = [
        #[1,2,3,4],
        #[1,1,1,2,3,3,4,5,5,5],
        [1,2,3,4,5],
    ]
    for case in cases:
        print(case)
        print(Solution().productExceptSelf(case))

